scripts/templates/valddi.cpp.mako

```diff
@@ -70,6 +70,40 @@ namespace ur_validation_layer
 
             %endfor
             %endfor
+
+            ## Handling stype validation gets a little intricate due to the
+            ## possibility of range and optional params. Note even this is
+            ## limited as it doesn't check structs recursively.
+            <% stype_params = th.get_stype_params(n, tags, obj, meta, specs) %>
+            %for param in stype_params:
+                <%
+                    param_name = param['name']
+                    if th.param_traits.is_range(param):
+                        param_name += "[i]"
+                    stype_check = "stype_map<{0}>::value != {1}->stype".format(
+                        param['type'], param_name)
+                    if th.param_traits.is_optional(param) and not th.param_traits.is_range(param):
+                        stype_check = param_name + " && " + stype_check
+                %>
+                %if th.param_traits.is_range(param):
```

Comment:
I don't see anything in `ur_valddi.cpp` that was generated using the code below.

Reply:
I thought at least [urProgramSetSpecializationConstants](https://oneapi-src.github.io/unified-runtime/core/api.html#urprogramsetspecializationconstants) would trigger it but `ur_specialization_constant_info_t` doesn't have a base struct, so no stype. I would argue that it probably should have a base, but I'd rather split that change into a separate PR. Testing with it actually generating stuff locally I found that some of the template was wrong so I've pushed the fix for that